classes/property_list.php

```
@@ -0,0 +1,346 @@
+     *
+     * @param string $key Key of element.
+     * @return mixed Value of element found, or null if none found.
+     */
+    public function get_element_value(string $key) {
+        $result = null;
+        $this->plist_map( function($elvalue, $elkey, $parent) use ($key, &$result) {
+            // Convert date to iso 8601 if date object.
+            if ($key === $elkey) {
+                $result = $elvalue->getValue();
+            }
+        }, $this->cfpropertylist->getValue());
+
+        if (is_array($result)) {
+            // Turn CFType elements in PHP elements.
+            $result = $this->array_serialize_cftypes($result);
+        }
+        return $result;
+    }
+
+    /**
+     * Update the value of any element with matching key.
+     *
+     * Only allow string, number and boolean elements to be updated.
+     *
+     * @param string $key Key of element to update.
+     * @param mixed $value  Value to update element with.
+     */
+    public function update_element_value(string $key, $value) {
+        if (is_array($value)) {
+            debugging('property_list: Can\'t call update_element_value with array. Use update_element_array instead.',
+                    DEBUG_DEVELOPER);
+            return;
+        }
+        $this->plist_map( function($elvalue, $elkey, $parent) use ($key, $value) {
+            // Set new value.
+            if ($key === $elkey) {
+                $element = $parent->get($elkey);
+                // Limit update to boolean and strings types, and check value matches expected type.
+                if (($element instanceof CFString && is_string($value))
+                        || ($element instanceof CFNumber && is_numeric($value))
+                        || ($element instanceof CFBoolean && is_bool($value))) {
+                    $element->setValue($value);
+                } else {
+                    debugging('property_list: Can only update strings, numbers and bools with corresponding type.',
```

Should we terminate and throw an exception and not silently swallow broken value?

---

Replaced debugging with exceptions for bad params if updating. Adjusted unit tests to expect exceptions instead of debugging.